analytics/uptime_data.go

```diff
@@ -1,20 +1,333 @@
+			}
+		}
+	}
+
+	fields := structs.Fields(a.Counter)
+	for _, field := range fields {
+		colName := "counter_" + field.Tag("json")
+
+		switch {
+		case strings.Contains(colName, "hits"), strings.Contains(colName, "error"), strings.Contains(colName, "success"):
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr("`" + colName + "` + " + fmt.Sprint(field.Value()))
+			}
+		case strings.Contains(colName, "total_request_time"):
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr("`" + colName + "` + " + fmt.Sprint(a.TotalRequestTime))
+			}
+		case strings.Contains(colName, "request_time"):
+			//AVG adding value to another AVG: newAve = ((oldAve*oldNumPoints) + x)/(oldNumPoints+1)
+			if !field.IsZero() {
+				assignments[colName] = gorm.Expr("((`" + colName + "` * `counter_hits`) + " + fmt.Sprint(a.RequestTime) + ")/( `counter_hits` +1)")
+			}
+		}
+
+	}
+
+	return assignments
+}
+
+func (u *UptimeReportAggregate) Dimensions() (dimensions []Dimension) {
+	for key, inc := range u.URL {
+		dimensions = append(dimensions, Dimension{"url", key, inc})
+	}
+
+	for key, inc := range u.Errors {
+		dimensions = append(dimensions, Dimension{"errors", key, inc})
+	}
+
+	dimensions = append(dimensions, Dimension{"", "total", &u.Total})
+
+	return
+}
+
+func (a *UptimeReportAggregateSQL) TableName() string {
+	return "tyk_uptime_analytics"
+}
+
+func (a *UptimeReportAggregateSQL) ProcessStatusCodes() {
+	for k, v := range a.Counter.ErrorMap {
+		switch k {
+		case "200":
+			a.Code200 = v
+		case "201":
+			a.Code201 = v
+		case "301":
+			a.Code301 = v
+		case "302":
+			a.Code302 = v
+		case "303":
+			a.Code303 = v
+		case "400":
+			a.Code400 = v
+		case "401":
+			a.Code401 = v
+		case "403":
+			a.Code403 = v
+		case "404":
+			a.Code404 = v
+		case "429":
+			a.Code429 = v
+		case "500":
+			a.Code500 = v
+		case "501":
+			a.Code501 = v
+		case "502":
+			a.Code502 = v
+		case "503":
+			a.Code503 = v
+		case "504":
+			a.Code504 = v
+		default:
+			switch k[0] {
+			case '1':
+				a.Code1x = v
+			case '2':
+				a.Code2x = v
+			case '3':
+				a.Code3x = v
+			case '4':
+				a.Code4x = v
+			case '5':
+				a.Code5x = v
+			}
+		}
+	}
+
+	a.Counter.ErrorList = nil
+	a.Counter.ErrorMap = nil
+}
+
+type UptimeReportAggregate struct {
+	TimeStamp time.Time
+	OrgID     string
+	TimeID    struct {
+		Year  int
+		Month int
+		Day   int
+		Hour  int
+	}
+
+	URL    map[string]*Counter
+	Errors map[string]*Counter
+
+	Total Counter
+
+	ExpireAt time.Time `bson:"expireAt" json:"expireAt"`
+	LastTime time.Time
+}
+
+func (u UptimeReportAggregate) New() UptimeReportAggregate {
+	agg := UptimeReportAggregate{}
+
+	agg.URL = make(map[string]*Counter)
+	agg.Errors = make(map[string]*Counter)
+
+	return agg
+}
+
+func AggregateUptimeData(data []UptimeReportData) map[string]UptimeReportAggregate {
+	analyticsPerOrg := make(map[string]UptimeReportAggregate)
+
+	for _, v := range data {
+		thisV := v
+		orgID := thisV.OrgID
```

Comment:
out of curiosity any reason why you assign `v` to `thisV`? since it's in same scope it's not necesary you can in for signature say 

```
for _, thisV := range data {
```

really minor thing, not merge stopper :)

Reply:
solved! thanks